File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const n = s.length;
    let i=0;
    let j=0;
    const mp = new Map();
    let maxlength=0;

    while(j<n){

        while(mp.has(s[j])){
            mp.set(s[i], mp.get(s[i])-1);
            if(mp.get(s[i])==0){
                mp.delete(s[i]);
            }
            i++;
        }

        mp.set(s[j],1);
        maxlength= Math.max(maxlength, j-i+1);
        j++;


    }
    return maxlength;
    
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        if(n<2) return n;

        int i=0;
        int j=0;
        int maxlength =0;
        map<int, int>mymap;

        while(j<n){

            while(mymap.find(s[j])!=mymap.end()){
                mymap[s[i]]--;
                if(mymap[s[i]]==0){
                    mymap.erase(s[i]);
                }
                i++;
            }
            mymap[s[j]]++;
            maxlength= max(maxlength, j-i+1);
            j++;

        }
        return maxlength;
    }
};